flink-streaming-java/src/main/java/org/apache/flink/streaming/util/retryable/AsyncRetryStrategies.java

```diff
@@ -0,0 +1,260 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.util.retryable;
+
+import org.apache.flink.streaming.api.functions.async.AsyncRetryPredicate;
+import org.apache.flink.streaming.api.functions.async.AsyncRetryStrategy;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/** Utility class to create concrete {@link AsyncRetryStrategy}. */
+public class AsyncRetryStrategies {
+    public static final NoRetryStrategy NO_RETRY_STRATEGY = new NoRetryStrategy();
+
+    /** NoRetryStrategy. */
+    public static class NoRetryStrategy implements AsyncRetryStrategy {
+        private static final long serialVersionUID = 1L;
+
+        private NoRetryStrategy() {}
+
+        @Override
+        public boolean canRetry(int currentAttempts) {
+            return false;
+        }
+
+        @Override
+        public long getBackoffTimeMillis(int currentAttempts) {
+            return -1;
+        }
+
+        @Override
+        public AsyncRetryPredicate getRetryPredicate() {
+            return new AsyncRetryPredicate() {
+                @Override
+                public Optional<Predicate<Collection>> resultPredicate() {
+                    return Optional.empty();
+                }
+
+                @Override
+                public Optional<Predicate<Throwable>> exceptionPredicate() {
+                    return Optional.empty();
+                }
+            };
+        }
+    }
+
+    /** FixedDelayRetryStrategy. */
+    public static class FixedDelayRetryStrategy<OUT> implements AsyncRetryStrategy<OUT> {
+        private static final long serialVersionUID = 1L;
+        private final int maxAttempts;
+        private final long backoffTimeMillis;
+        private final Predicate<Collection<OUT>> resultPredicate;
+        private final Predicate<Throwable> exceptionPredicate;
+
+        private FixedDelayRetryStrategy(
+                int maxAttempts,
+                long backoffTimeMillis,
+                Predicate<Collection<OUT>> resultPredicate,
+                Predicate<Throwable> exceptionPredicate) {
+            this.maxAttempts = maxAttempts;
+            this.backoffTimeMillis = backoffTimeMillis;
+            this.resultPredicate = resultPredicate;
+            this.exceptionPredicate = exceptionPredicate;
+        }
+
+        @Override
+        public boolean canRetry(int currentAttempts) {
+            return currentAttempts <= maxAttempts;
+        }
+
+        @Override
+        public AsyncRetryPredicate<OUT> getRetryPredicate() {
+            return new AsyncRetryPredicate<OUT>() {
+                @Override
+                public Optional<Predicate<Collection<OUT>>> resultPredicate() {
```

Comment:
Might be simplified with `return Optional.ofNullable(resultPredicate);`. The same to the other methods. 

Also if not consider the possible changes to the API, I think we could extract a named inner `AsyncRetryPredicate` class that composes of two separate predicates.